pkg/customunit/manager.go

```diff
@@ -0,0 +1,331 @@
+package customunit
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/Axway/agent-sdk/pkg/agent/cache"
+	"github.com/Axway/agent-sdk/pkg/amplify/agent/customunits"
+	v1 "github.com/Axway/agent-sdk/pkg/apic/apiserver/models/api/v1"
+	"github.com/Axway/agent-sdk/pkg/apic/apiserver/models/catalog/v1"
+	management "github.com/Axway/agent-sdk/pkg/apic/apiserver/models/management/v1alpha1"
+	"github.com/Axway/agent-sdk/pkg/apic/definitions"
+	"github.com/Axway/agent-sdk/pkg/config"
+	"github.com/Axway/agent-sdk/pkg/transaction/models"
+	transUtil "github.com/Axway/agent-sdk/pkg/transaction/util"
+	"github.com/Axway/agent-sdk/pkg/util"
+	"github.com/Axway/agent-sdk/pkg/util/log"
+)
+
+type CustomUnitMetricServerManager struct {
+	configs          []config.MetricServiceConfiguration
+	cache            cache.Manager
+	agentType        config.AgentType
+	logger           log.FieldLogger
+	clients          []*customUnitClient
+	metricReportChan chan *customunits.MetricReport
+	stopChan         chan struct{}
+}
+
+type metricCollector interface {
+	AddCustomMetricDetail(models.CustomMetricDetail)
+}
+
+func NewCustomUnitMetricServerManager(configs []config.MetricServiceConfiguration, cache cache.Manager, agentType config.AgentType) *CustomUnitMetricServerManager {
+	return &CustomUnitMetricServerManager{
+		configs:          configs,
+		cache:            cache,
+		agentType:        agentType,
+		metricReportChan: make(chan *customunits.MetricReport, 100),
+		stopChan:         make(chan struct{}),
+		clients:          []*customUnitClient{},
+		logger:           log.NewFieldLogger().WithPackage("customunit").WithComponent("manager"),
+	}
+}
+
+func (h *CustomUnitMetricServerManager) HandleQuotaEnforcement(ctx context.Context, cancelCtx context.CancelFunc, ar *management.AccessRequest, app *management.ManagedApplication) error {
+	// Build quota info
+	quotaInfo, err := h.buildQuotaInfo(ctx, ar, app)
+	if err != nil {
+		return fmt.Errorf("could not build quota info from access request")
+	}
+	errMessage := ""
+	for _, config := range h.configs {
+		if config.MetricServiceEnabled() {
+			factory := NewCustomUnitClientFactory(config.URL, h.cache, quotaInfo)
```

Comment:
We should have the factories created in the constructor with things that do not change on each connection, such as the cache that is sent in and any grpc opts that are used. Then the factory will receive only the args that differ between services or calls, url, and quotaInfo here.

Similar for the metric service factory below